### PR TITLE
Add GCN-BMP

### PR DESCRIPTION
**Pull request by mughetto**
# Summary
 
First draft of implementation for GCN-BMP. Closes Issue #21 .

- [ ] Unit tests provided for these changes
- [x] Documentation and docstrings added for these changes

## Changes 

* Add GCN-BMP model. Some parts might be moved in other files/projects:
   - Circular correlation function.
   - Highway update
   - Attention-based pooling/readout
* Add example for the model.

## Caveats
* The number of relationships (for the Relational Conv layers) is configured in the model initialization. It should be done via some dataset-level constant.

## Differences with the original paper
* In the original GCN-BMP the feature vectors are initialized randomly based on the atomic number of the node. I couldn't find easily the implementation in their code so backed up on the features already provided in the PairedBatches.
* Rel Conv layers are followed by a sigmoid activation before the Highway Update. This should only affect the stability.
* Highway Update used is the one proposed in the original Highway network paper. Best for reusability imho and shouldn't affect the gating behavour.

## Performance
Example's ROC AUC: 0.749128

### CODE REVIEW

File: chemicalx/models/gcnbmp.py [hunk 1]
```diff
@@ -1,14 +1,257 @@
...
+        """
+        features_left = self.graph_convolutions(molecules_left, molecules_left.data_dict["node_feature"])[
+            "node_feature"
+        ]
+
+        features_right = self.graph_convolutions(molecules_right, molecules_right.data_dict["node_feature"])[
+            "node_feature"
+        ]
+
+        features_left = self.attention_readout_left(
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:246]: Attention should be symmetric. The same instance is used on both sides.

File: chemicalx/models/gcnbmp.py [hunk 2]
```diff
@@ -1,14 +1,257 @@
...
+        if self.concat_hidden:
+            node_feature = torch.cat(hiddens, dim=-1)
+        else:
+            node_feature = hiddens[-1]
+
+        return {"node_feature": node_feature}
+
+
+class GCNBMP(Model):
+    r"""An implementation of the GCNBMP model.
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:189]: Model reference - see other model class docstrings.

File: chemicalx/models/gcnbmp.py [hunk 3]
```diff
@@ -1,14 +1,257 @@
 """An implementation of the GCNBMP model."""
+import sys
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:2]: Check unused imports.

File: chemicalx/models/gcnbmp.py [hunk 4]
```diff
@@ -1,14 +1,257 @@
...
-from .base import UnimplementedModel
 
 __all__ = [
     "GCNBMP",
 ]
 
 
-class GCNBMP(UnimplementedModel):
-    """An implementation of the GCNBMP model.
+def circular_correlation(left_x, right_x):
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:26]: Input types and return types - extra space.

File: chemicalx/models/gcnbmp.py [hunk 5]
```diff
@@ -1,14 +1,257 @@
...
+    :return circ_corr.real (Tensor): joint representation by circular correlation.
+    """
+    left_x_cfft = torch.conj(fft(left_x))
+    right_x_fft = fft(right_x)
+    circ_corr = ifft(torch.mul(left_x_cfft, right_x_fft))
+
+    return circ_corr.real
+
+
+class Highway(nn.Module):
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:41]: Input types and return types everywhere.

File: chemicalx/models/gcnbmp.py [hunk 6]
```diff
@@ -1,14 +1,257 @@
...
+        :param prev_input_size (int): representation obtained by the previous convolutional layer.
+        """
+        super(Highway, self).__init__()
+        total_size = input_size + prev_input_size
+        self.proj = nn.Linear(total_size, input_size)
+        self.transform = nn.Linear(total_size, input_size)
+        # Not in GCN-BMP paper but present in the original implementation
+        self.transform.bias.data.fill_(-2.0)
+
+    def forward(self, input, prev_input):
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:54]: Typing.

File: chemicalx/models/gcnbmp.py [hunk 7]
```diff
@@ -1,14 +1,257 @@
...
+        self.layers = nn.ModuleList()
+        for i in range(len(self.dims) - 1):
+            self.layers.append(
+                layers.RelationalGraphConv(
+                    self.dims[i], self.dims[i + 1], num_relation, edge_input_dim, batch_norm, activation
+                )
+            )
+            self.layers.append(Highway(self.dims[i + 1], self.dims[i]))
+
+    def forward(self, graph, input):
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:145]: Input typing and returns.

File: chemicalx/models/gcnbmp.py [hunk 8]
```diff
@@ -1,14 +1,257 @@
...
+        Returns:
+            g (Tensor): graph-level representation
+        """
+        att = torch.sigmoid(self.lin(torch.cat((input_rep, final_rep), 1)))
+        g = att.mul(self.last_rep(final_rep))
+        g = scatter_add(g, graph_index, dim=0)
+        return g
+
+
+class GCNBMPEncoder(nn.Module, core.Configurable):
```
Comment (benedekrozemberczki) [inline chemicalx/models/gcnbmp.py:104]: Input typings everywhere.

**Review by benedekrozemberczki — CHANGES_REQUESTED** [review]
Generally looks good, typing is missing in certain places. The auxiliary layers should be in the readme on some name space - also you should consider adding them to the README.

File: chemicalx/models/gcnbmp.py (hunk 1 above)
Comment (benedekrozemberczki) [reply chemicalx/models/gcnbmp.py:246]: This overparametrizes the model.

Comment (mughetto) [reply chemicalx/models/gcnbmp.py:244]: Fixed, in the same spirit note that the circular correlation operator is not symmetrical un L<->R permutation. It was managed by the authors by duplicating each pair of drugs.

File: chemicalx/models/gcnbmp.py [hunk 9]
```diff
@@ -1,14 +1,255 @@
...
+            hidden_dims = [hidden_dims]
+        self.input_dim = input_dim
+        self.output_dim = hidden_dims[-1] * (len(hidden_dims) if concat_hidden else 1)
+        self.dims = [input_dim] + list(hidden_dims)
+        self.num_relation = num_relation
+        self.short_cut = short_cut
+        self.concat_hidden = concat_hidden
+
+        self.layers = nn.ModuleList()
+        for i in range(len(self.dims) - 1):
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:135]: use `from more_itertools import pairwise` and then do

```suggestion
        for left_dim, right_dim in pairwise(self.dims):
```

File: chemicalx/models/gcnbmp.py [hunk 10]
```diff
@@ -1,14 +1,255 @@
...
+
+        Returns:
+            dict with ``node_feature`` field:
+                node representations of shape :math:`(|V|, d)`
+        """
+        hiddens = []
+        layer_input = input
+        prev_gcn = input
+
+        for i, layer in enumerate(self.layers):
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:161]: use `from more_itertools import chunked` to do both operations together
```suggestion
       for conv, highway in chunked(self.layers, 2):
```

File: chemicalx/models/gcnbmp.py [hunk 11]
```diff
@@ -37,10 +38,10 @@ def circular_correlation(left_x, right_x):
 
 
 class Highway(nn.Module):
-    def __init__(self, input_size, prev_input_size):
+    def __init__(self, input_size: int, prev_input_size: int):
         """Instantiate the Highway update layer based on https://arxiv.org/abs/1505.00387
-        :param input_size (int): current representation.
-        :param prev_input_size (int): representation obtained by the previous convolutional layer.
+        :param input_size (int): current representation size.
+        :param prev_input_size (int): size of the representation obtained by the previous convolutional layer.
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:44]: while you're adding the python type annotations, please take them out of the docstrings (they're not only redundant but also noisy when trying to read)

File: chemicalx/models/gcnbmp.py (hunk 10 above)
Comment (mughetto) [reply chemicalx/models/gcnbmp.py:162]: Thanks, didn't know about this one. Will also add it to the dependencies

File: chemicalx/models/gcnbmp.py (hunk 11 above)
Comment (mughetto) [reply chemicalx/models/gcnbmp.py:44]: Ok I was tryingto follow what I was seeing in some other models, will clean it now.

File: chemicalx/models/gcnbmp.py [hunk 12]
```diff
@@ -1,14 +1,255 @@
...
+        ]
+
+        features_right = self.graph_convolutions(molecules_right, molecules_right.data_dict["node_feature"])[
+            "node_feature"
+        ]
+
+        features_left = self.attention_readout(
+            molecules_left.data_dict["node_feature"], features_left, molecules_left.node2graph
+        )
+        features_right = self.attention_readout(
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:247]: note that there's now code duplication in the application of the graph convolutions and the attention readout to the left and right features. Please make a helper function to reduce redundant code

File: examples/gcnbmp_examples.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+"""Example with EPGCNDS."""
```
Comment (cthoyt) [inline examples/gcnbmp_examples.py:1]: check name

File: chemicalx/models/gcnbmp.py [hunk 14]
```diff
@@ -1,17 +1,263 @@
...
+        return {"node_feature": node_feature}
 
-class GCNBMP(UnimplementedModel):
-    """An implementation of the GCN-BMP model from [chen2020]_.
 
-    .. seealso:: This model was suggested in https://github.com/AstraZeneca/chemicalx/issues/21
+class GCNBMP(Model):
+    r"""The GCNBMP model from [gcnbmp].
 
-    .. [chen2020] Chen, X., *et al.* (2020). `GCN-BMP: Investigating graph representation learning
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:15]: please be careful when merging - the citation on the main branch is the desired one.

File: chemicalx/models/gcnbmp.py [hunk 15]
```diff
@@ -1,17 +1,265 @@
...
+        self.lin = nn.Linear(
+            total_features_channels, hidden_channels
+        )  # weights here must be shared across all nodes according to the paper
+        self.last_rep = nn.Linear(hidden_channels, hidden_channels)
+
+    def forward(self, input_rep: torch.Tensor, final_rep: torch.Tensor, graph_index: torch.Tensor) -> torch.Tensor:
+        """
+        Compute an attention-based readout using the input and output layers of the RGCN encoder for one molecule.
+
+        Parameters:
```
Comment (cthoyt) [inline chemicalx/models/gcnbmp.py:104]: switch to sphinx-style docs

**Comment by benedekrozemberczki** [comment]
@cthoyt I overtook this.

**Comment by cthoyt** [comment]
@benedekrozemberczki let me know when you're done and I'll take a final pass

**Comment by benedekrozemberczki** [comment]
@cthoyt Ready for review.

**Review by benedekrozemberczki — APPROVED** [review]
